primitives/runtime/src/offchain/storage.rs

```diff
@@ -58,30 +77,44 @@ impl<'a> StorageValueRef<'a> {
 	/// Retrieve & decode the value from storage.
 	///
 	/// Note that if you want to do some checks based on the value
-	/// and write changes after that you should rather be using `mutate`.
+	/// and write changes after that, you should rather be using `mutate`.
 	///
-	/// The function returns `None` if the value was not found in storage,
-	/// otherwise a decoding of the value to requested type.
-	pub fn get<T: codec::Decode>(&self) -> Option<Option<T>> {
-		sp_io::offchain::local_storage_get(self.kind, self.key)
-			.map(|val| T::decode(&mut &*val).ok())
+	/// Returns an error if the value was not found in storage
+	/// - either because the key didn't exist or the value could not be decoded.
+	/// Otherwise returns a decoding of the value to requested type.
+	pub fn get<T: codec::Decode>(&self) -> Result<T, StorageRetrivalError> {
+		let val = sp_io::offchain::local_storage_get(self.kind, self.key)
+			.ok_or(StorageRetrivalError::NoKey)?;
+
+		T::decode(&mut &*val).map_err(|_|{StorageRetrivalError::Undecodable})
 	}
 
-	/// Retrieve & decode the value and set it to a new one atomically.
+	/// Retrieve & decode the current value and set it to a new value atomically.
+	///
+	/// Function `mutate_val` takes as input the current value and should
+	/// return a new value that is attempted to be written to storage.
 	///
-	/// Function `f` should return a new value that we should attempt to write to storage.
 	/// This function returns:
-	/// 1. `Ok(Ok(T))` in case the value has been successfully set.
-	/// 2. `Ok(Err(T))` in case the value was calculated by the passed closure `f`,
-	///    but it could not be stored.
-	/// 3. `Err(_)` in case `f` returns an error.
-	pub fn mutate<T, E, F>(&self, f: F) -> Result<Result<T, T>, E> where
+	/// 1. `Ok(T)` in case the value has been successfully set.
+	/// 2. `Err(MutateStorageError::CalculatedButNotStored(T))` in case the value was calculated
+	/// by the passed closure `mutate_val`, but it could not be stored.
+	/// 3. `Err(MutateStorageError::ValueFunctionFailed(_))` in case `mutate_val` returns an error.
+	pub fn mutate<T, E, F>(&self, mutate_val: F) -> Result<T, MutateStorageError<T,E>> where
 		T: codec::Codec,
-		F: FnOnce(Option<Option<T>>) -> Result<T, E>
+		F: FnOnce(Result<T, StorageRetrivalError>) -> Result<T, E>
 	{
 		let value = sp_io::offchain::local_storage_get(self.kind, self.key);
-		let decoded = value.as_deref().map(|mut v| T::decode(&mut v).ok());
-		let val = f(decoded)?;
+		let decoded =
+			match value.as_deref() {
+				Some(mut v) => {
+					T::decode(&mut v).map_err(|_| {StorageRetrivalError::Undecodable })
```

Comment:
Same as above: you can simplify into

```rust
let bytes = value.as_deref().ok_or(StorageRetrievalError::Nokey)?;
T::decode(&mut bytes).map_err(...)
```

Reply:
decode doesn't take a Result. I'm wondering what value.as_deref() into because Input is not defined on a mutable slice which decode takes... - I was pitcing for `.and_then` but keeps telling me bytes isn't Input.

Reply:
(Doesn't help that I've also hit an ICE but at least there is already a PR for that)

Reply:
ah I get it now: It's a &mut of a &[u8] and so is `mut` and `Input`. Hmm, not sure what I think about that.